### PR TITLE
GroupedHashAggregateStream should register spillable consumer

### LINKED ISSUE #8001: GroupedHashAggregateStream should register spillable consumer
### Describe the bug

`MemoryConsumer` for `GroupedHashAggregateStream` should be `with_can_spill=true` 

change from:

https://github.com/apache/arrow-datafusion/blob/a9d66e2b492843c2fb335a7dfe27fed073629b09/datafusion/physical-plan/src/aggregates/row_hash.rs#L331

to: 

```rust
let reservation = MemoryConsumer::new(name).with_can_spill(true).register(context.memory_pool());
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

Relates to #7858

### PR DESCRIPTION
**Pull request by milenkovicm**
## Which issue does this PR close?

Closes #8001.

## Rationale for this change

It is expected for this operator to register spillable consumer as it can spill

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/physical-plan/src/aggregates/row_hash.rs [hunk 1]
```diff
@@ -328,7 +328,9 @@ impl GroupedHashAggregateStream {
             .collect();
 
         let name = format!("GroupedHashAggregateStream[{partition}]");
-        let reservation = MemoryConsumer::new(name).register(context.memory_pool());
+        let reservation = MemoryConsumer::new(name)
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/row_hash.rs:331]: https://docs.rs/datafusion/latest/datafusion/execution/memory_pool/struct.MemoryConsumer.html#method.with_can_spill

I agree 

@kazuyukitanimura and @yjshen  do you concur? Is there any way we could potentially test this behavior?

**Review by alamb — APPROVED** [review]
Thank you @milenkovicm

File: datafusion/physical-plan/src/aggregates/row_hash.rs (hunk 1 above)
Comment (milenkovicm) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:331]: Can we create TestMemoryPool which will panic if unspillable operator is registered ?

Comment (kazuyukitanimura) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:331]: Yes, looks good.
Another idea to test is that to write a test aggregation job with small memory pool size that fails with the greedy pool but succeeds with the fair pool. Not sure if we can write such case...

**Review by kazuyukitanimura — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Pull request merged**